crates/adapters/src/transport/durable/kafka/input.rs

```diff
@@ -0,0 +1,771 @@
+            ("retries", "5"),
+            ("socket.nagle.disable", "true"),
+            ("linger.ms", "0"),
+        ];
+        let mut producer_config = Self::kafka_config(
+            &source,
+            &source.producer_options,
+            PRODUCER_SETTINGS,
+            "producer",
+        )?;
+        producer_config.remove("group.id");
+
+        let index_topic = match source.index_topic {
+            Some(ref s) => s.clone(),
+            None => format!("{}_input-index", source.data_topic),
+        };
+
+        Ok(Inner {
+            index_consumer_config,
+            reader_consumer_config,
+            producer_config,
+            data_topic: source.data_topic.clone(),
+            index_topic,
+            partition: source.partition,
+        })
+    }
+}
+
+/// Durable Kafka input endpoint.
+struct Endpoint(Inner);
+
+/// Data type stored in a step partition, to associate a DBSP step number with
+/// the corresponding range of event offsets in the data partition.
+#[derive(Deserialize, Serialize, Clone, Debug, PartialEq, Eq, Hash)]
+struct IndexEntry {
+    step: Step,
+    #[serde(serialize_with = "serialize_range")]
+    data_offsets: Range<i64>,
+}
+
+impl IndexEntry {
+    fn read(ctp: &Ctp, offset: i64) -> AnyResult<IndexEntry> {
+        ctp.seek(offset)?;
+        ctp.resume()?;
+        let msg = ctp.consumer.poll(None).ok_or_else(|| {
+            anyhow!(
+                "partition {} in {} should not be empty",
+                ctp.partition,
+                ctp.topic
+            )
+        })??;
+        let payload = msg.payload().unwrap_or(&[]);
+        serde_json::from_slice::<IndexEntry>(payload).with_context(|| {
+            format!(
+                "message at offset {offset} in topic {} partition {} should be valid IndexEntry",
+                ctp.topic, ctp.partition
+            )
+        })
+    }
+}
+
+/// Serializes `r` as `[start, end]` rather than as `["start": start, "end":
+/// end]`.
+fn serialize_range<S: Serializer>(r: &Range<i64>, serializer: S) -> Result<S::Ok, S::Error> {
+    let mut tup = serializer.serialize_tuple(2)?;
+    tup.serialize_element(&r.start)?;
+    tup.serialize_element(&r.end)?;
+    tup.end()
+}
+
+#[cfg(test)]
+#[test]
+fn test_index_entry() {
+    let data = r#"
+{
+    "step": 123,
+    "data_offsets": [12, 23]
+}
+"#;
+    let entry: IndexEntry = serde_json::from_str(data).unwrap();
+    assert_eq!(
+        entry,
+        IndexEntry {
+            step: 123,
+            data_offsets: 12..23
+        }
+    );
+}
+
+/// Consumer, topic, and partition (in Kafka).
+///
+/// [`rdkafka`] uses these three pieces together for a lot of calls, without
+/// providing a type to bind them together.  This should help.
+struct Ctp<'a> {
+    consumer: &'a BaseConsumer<DefaultConsumerContext>,
+    topic: &'a str,
+    partition: i32,
+}
+impl<'a> Ctp<'a> {
+    fn new(
+        consumer: &'a BaseConsumer<DefaultConsumerContext>,
```

Comment:
It's not necessary for partition rebalance events, because we're assigning partitions to workers ourselves and we will never get those.

I'll figure things out about errors.

Reply:
I think there might still be an initial rebalance event when the consumer subscribes to the partition, and the final one on disconnect (or are you using the API in a way that disables those as well?) The latter is not an issue.  For the former, we want to make sure the following doesn't happen:
- the client creates pipeline in the PAUSED state
- the initial rebalance event arrives, subscribing the connector to the assigned partition
- the connector doesn't handle the rebalance event and therefore doesn't pause the partition.

Reply:
>I think there might still be an initial rebalance event when the consumer subscribes to the partition...

The Kafka documentation says that rebalancing, and rebalancing events, are only for consumers where we're not assigning partitions and positions ourselves. But the Kafka documentation is often wrong, so I'll add some assertions or log messages around this.

Reply:
I suspect Kafka documentation is correct in this case, and I am wrong.